Updated Development Guide

learnjavawithreshma opened the pull request:
Has the following changes:

- Setting up the code in Eclipse (Cloning the repository and creating Eclipse Project)

- Building the application and running it standalone

A Few other notes:

- In the database section, I am not sure if the steps for PostgreSQL are correct, I have written them based on the existing steps and some steps from the Getting Started Document. I can download and setup PostGreSQL and verify these steps if required. 

- For the Customizing Log path, I have taken the existing logback file and simple added an additional appender for an external log file. However, the existing logback file has many appenders related to JDBC. Can I remove the JDBC appenders from the logback file in the document in order to keep things simple?

- I'm yet to update the dependencies and debugging section. I will do that next.

learnjavawithreshma committed:
Updated Development Guide

Updated Development Guide

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -1,129 +1,530 @@
-----
-mvn clean install -DskipTests
-----
+. In the Git Repository details, enter the Kill Bill Git URL https://github.com/killbill/killbill in the URI field. Enter your Git Credentials. Click *Next*
```
Are credentials really required? I believe one could clone the repository anonymously.

learnjavawithreshma replied:
Updated document to include steps for cloning the repository anonymously

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -1,129 +1,530 @@
+. In the branch selection view, select the branch named *master* (By default, all the branches appear selected, you need to uncheck all branches except for *master*). Click *Next*
 
-Once built, you can start Kill Bill by running:
+. Select the path on your computer where you would like to clone the code. This path will be referred to as *PROJECT_ROOT* from now on).For example, if you choose `C:/MyProjects/KillBill`, *PROJECT_ROOT* refers to this path. Click *Finish* (This starts the cloning process and may take a while depending on your internet speed)
```
Missing space after `.`: `on).For`

learnjavawithreshma replied:
Fixed

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -1,129 +1,530 @@
+
+. Ensure that you have *Maven 3.5.2 or higher* (It can be downloaded from http://maven.apache.org/download.cgi[here])
+
+. Maven requires JDK. Ensure that you have *JDK 7 or higher* (It can be downloaded from
```
It's actually JDK 8 or higher now.

learnjavawithreshma replied:
Fixed

pierre reviewed (commented):
Looks great so far!

pierre commented:
> * In the database section, I am not sure if the steps for PostgreSQL are correct, I have written them based on the existing steps and some steps from the Getting Started Document. I can download and setup PostGreSQL and verify these steps if required.

If you could do that (download and verify the PostgreSQL steps), that would be great! Thanks!

> * For the Customizing Log path, I have taken the existing logback file and simple added an additional appender for an external log file. However, the existing logback file has many appenders related to JDBC. Can I remove the JDBC appenders from the logback file in the document in order to keep things simple?

Yup, you can safely remove these. These are disabled by default (see `level="OFF"`) and mostly useful to debug production issues.

> I'm yet to update the dependencies and debugging section. I will do that next.

👍

learnjavawithreshma committed:
Updated development doc

1) Fixed all comments
2) Added PostGreSQL set up steps
3) Added remote debugging steps
4) Added a new section - Some common errors and their solutions

learnjavawithreshma committed:
Created separate sections for MySQL and PostGreSQL

learnjavawithreshma committed:
Minor fixes

Minor fixes to log path section and common errors section

learnjavawithreshma commented:
Added the following additional stuff:

- Added PostGreSQL steps (Now I'm wondering if such detailed steps should be provided for MySQL as well)

- Added remote debugging steps 

- Added a new section - Some common errors and their solutions

I'm not sure what to do about the following:

- Dependencies section - I believe these are some additional dependencies which may need to be built.  I'm not sure how this fits into the Kill Bill development set up.  Can this section be removed or do we need to explain this better?

- There are some additional sections in the document like **Navigating the Kill Bill code base, Date, Time and Timezone, Adding a new Cache, Adding a new Plugin API, Plugin compatibility with API changes, Releases**. I may be wrong, but they look like they are not related to the development set up. So, should these sections be updated/removed from the development document?

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -20,11 +20,16 @@ The first step is to clone the Kill Bill repository. For this, you need to follo
-. In the Git Repository details, enter the Kill Bill Git URL https://github.com/killbill/killbill in the URI field. Enter your Git Credentials. Click *Next*
+. In the Git Repository details, enter the Kill Bill Git URL https://github.com/killbill/killbill in the *URI* field. 
+
+.. If you want read-write access to the Kill Bill repository, Select *https* in the *Protocol* dropdown and enter your Git credentials in the *User/Password* fields . 
```
Note: read-write access for the `github.com/killbill/killbill` repository is restricted to the members of the core team. However, people can fork the code to their own account (e.g. to `github.com/learnjavawithreshma/killbill`), and this repository would be read-write. I would clarify this.

learnjavawithreshma replied:
Updated as mentioned above

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -20,11 +20,16 @@ The first step is to clone the Kill Bill repository. For this, you need to follo
+. In the Git Repository details, enter the Kill Bill Git URL https://github.com/killbill/killbill in the *URI* field. 
+
+.. If you want read-write access to the Kill Bill repository, Select *https* in the *Protocol* dropdown and enter your Git credentials in the *User/Password* fields . 
+.. If you do not have Git credentials or want a read-only access, you can Select *Git* in the *Protocol* Dropdown. This disables the credentials fields. However, this will clone the repository anonymously and give you a read only access to the repository. 
```
Didn't you have to set up an SSH key associated with your GitHub account as well? Or maybe you did it initially when you set up the killbill-docs repo? If I go to https://github.com/killbill/killbill while being logged-out, I don't see the Git protocol in the dropdown for instance.

learnjavawithreshma replied:
No, I did not do anything special for setting up SSH key.  If required, I can read up a bit more about this and see why/how it worked for me.

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -130,39 +132,64 @@ Once the repository is cloned, you need to create an Eclipse project. For this,
 </lifecycleMappingMetadata>
 
-. Go to `Eclipse Preferences > Maven > Lifecycle Mappings`.Click *Browse* and select the path of the `lifecycle-mapping-metadata.xml` file
+. Go to `Window > Preferences > Maven > Lifecycle Mappings`.Click *Browse* and select the path of the `lifecycle-mapping-metadata.xml` file. Click `Apply and Close`
```
Missing space before `.`.

learnjavawithreshma replied:
Fixed

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -130,39 +132,64 @@ Once the repository is cloned, you need to create an Eclipse project. For this,
+==== PostGreSQL configuration
 
-.. Create a killbill user and database (verify connectivity with `psql -h 127.0.0.1 -U killbill -p 5432`)
+In order to configure PostGreSQL, you need to follow the steps given below (These steps can be executed either via the command line *psql* tool or via *pgAdmin*):
```
`PostgreSQL`

learnjavawithreshma replied:
Fixed

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -130,39 +132,64 @@ Once the repository is cloned, you need to create an Eclipse project. For this,
+
+. Execute https://github.com/killbill/killbill/blob/master/util/src/main/resources/org/killbill/billing/util/ddl-postgresql.sql[this] schema extension DDL. 
+
+. Run the Kill Bill DDL https://docs.killbill.io/latest/ddl.sql[here].In psql, you  can run the following command:
```
Space before `.`

learnjavawithreshma replied:
Fixed

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -280,14 +307,15 @@ Once the build is successful, you can run the application to verify that everyth
 . Modify the `PROJECT_ROOT/profiles/killbill/src/main/resources/killbill-server.properties`. Update the following properties as per your database:
 
- org.killbill.dao.url=jdbc:mysql://127.0.0.1:3306/killbill
+ org.killbill.dao.url=jdbc:mysql://127.0.0.1:3306/killbill or jdbc:postgresql://127.0.0.1:5432/killbill?currentSchema=killbillschema
```
I would add a comment instead:

```
# For PostgreSQL, use: jdbc:postgresql://127.0.0.1:5432/killbill?currentSchema=killbillschema
org.killbill.dao.url=jdbc:mysql://127.0.0.1:3306/killbill
```

learnjavawithreshma replied:
Changed

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -306,12 +334,12 @@ Once the build is successful, you can run the application to verify that everyth
+  <configuration>
    <jmxConfigurator />
-   <property name="LOGS_DIR" value="F:/killbill/logs" />
+   <property name="LOGS_DIR" value="<log_file_path>" />
```
Maybe mention above that `<log_file_path>` needs to be updated

learnjavawithreshma replied:
This is already mentioned in the subsequent step

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -507,24 +375,129 @@ The steps above configure the application so that the Kill Bill logs are display
+
+... Click *Attach Source*. 
+
+... Click on *External Location* and Navigate to your Maven local repository (`~/.m2/repository`). Select `/org/kill-bill/billing/killbill-platform-base/0.40.4/killbill-platform-base-0.40.4-sources.jar`
```
Maybe specify that this would change depending on the file that needs to be debugged (the version might change too).

learnjavawithreshma replied:
Fixed the description to make it more general. Please let me know if you still feel that some additional information is required.

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -507,24 +375,129 @@ The steps above configure the application so that the Kill Bill logs are display
+ [ERROR] After correcting the problems, you can resume the build with the command
+ [ERROR]   mvn <args> -rf :killbill-profiles-killbill
+ 
+This error is most probably because the `logback.xml` file is missing license information. This might have got deleted while customizing the log file path. So ensure that the `logback.xml` file has the following license information at the top:
```
You can actually specify `-Dcheck.skip-rat=true` on the command line. Might be easier to do so than adding the license back.

learnjavawithreshma replied:
Changed

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -507,24 +375,129 @@ The steps above configure the application so that the Kill Bill logs are display
+
+. This runs the application in debug mode. so, you can set additional breakpoints and debug as required.
+
+=== Some common errors and their solutions
```
Very cool section! 👍

pierre reviewed (changes requested):
👏 
Just a few suggestions and typos.

pierre commented:
> * Added PostGreSQL steps (Now I'm wondering if such detailed steps should be provided for MySQL as well)

👍  Feel free to expand the existing docs where you think this is needed.

> * Dependencies section - I believe these are some additional dependencies which may need to be built.  I'm not sure how this fits into the Kill Bill development set up.  Can this section be removed or do we need to explain this better?

> * There are some additional sections in the document like **Navigating the Kill Bill code base, Date, Time and Timezone**

I think we should keep these sections, as they explain:

1. Where to find the code (there are more than one GitHub repository)
2. What to look for as a developer when manipulating dates

> **Adding a new Cache, Adding a new Plugin API, Plugin compatibility with API changes, Releases**. I may be wrong, but they look like they are not related to the development set up. So, should these sections be updated/removed from the development document?

Agreed, I've update the doc here: https://github.com/killbill/killbill-docs/commit/ad6fd7181715c0f1e03058b1b7d96bfaceac8bf2.

pierre commented on userguide/tutorials/development.adoc:
```diff
@@ -1,127 +1,501 @@
+
+    create database killbill;
+   
+. Run the Kill Bill DDL https://docs.killbill.io/latest/ddl.sql[here].
```
As discussed on Slack, let's add more step-by-step instructions here as well for MySQL.

learnjavawithreshma replied:
Done

learnjavawithreshma committed:
Update development.adoc

learnjavawithreshma commented:
> > 👍 Feel free to expand the existing docs where you think this is needed.
>  
> Done, expanded the MySQL section

>> I think we should keep these sections, as they explain:
> >
> >1. Where to find the code (there are more than one GitHub repository)
> >2. What to look for as a developer when manipulating dates
> 
> Okay, I've left the **Dependencies, Navigating the Kill Bill code base, Date, Time and Timezone** sections as they are

>> Agreed, I've update the doc here: [ad6fd71](https://github.com/killbill/killbill-docs/commit/ad6fd7181715c0f1e03058b1b7d96bfaceac8bf2).
>
> Okay, so from what I understand, you've removed the **Adding a new Cache, Adding a new Plugin API, Plugin compatibility with API changes, Releases** sections from the original document. Is that correct? So, when my changes are merged with the original document will this be reflected?

learnjavawithreshma committed:
Update development.adoc

learnjavawithreshma committed:
Update development.adoc
